commit 6d157e7cb8d3f4aa28a9fef95c498ce8ac3c31a9
Author: Tom Lane <tgl@sss.pgh.pa.us>
Date:   Mon Jun 6 11:20:21 2022 -0400

    Don't fail on libpq-generated error reports in ecpg_raise_backend().
    
    An error PGresult generated by libpq itself, such as a report of
    connection loss, won't have broken-down error fields.
    ecpg_raise_backend() blithely assumed that PG_DIAG_MESSAGE_PRIMARY
    would always be present, and would end up passing a NULL string
    pointer to snprintf when it isn't.  That would typically crash
    before 3779ac62d, and it would fail to provide a useful error report
    in any case.  Best practice is to substitute PQerrorMessage(conn)
    in such cases, so do that.
    
    Per bug #17421 from Masayuki Hirose.  Back-patch to all supported
    branches.
    
    Discussion: https://postgr.es/m/17421-790ff887e3188874@postgresql.org

diff --git a/src/interfaces/ecpg/ecpglib/error.c b/src/interfaces/ecpg/ecpglib/error.c
index cd6c6a6819b..26fdcdb69e9 100644
--- a/src/interfaces/ecpg/ecpglib/error.c
+++ b/src/interfaces/ecpg/ecpglib/error.c
@@ -227,22 +227,21 @@ ecpg_raise_backend(int line, PGresult *result, PGconn *conn, int compat)
 		ecpg_log("out of memory");
 		ECPGfree_auto_mem();
 		return;
 	}
 
-	if (result)
-	{
-		sqlstate = PQresultErrorField(result, PG_DIAG_SQLSTATE);
-		if (sqlstate == NULL)
-			sqlstate = ECPG_SQLSTATE_ECPG_INTERNAL_ERROR;
-		message = PQresultErrorField(result, PG_DIAG_MESSAGE_PRIMARY);
-	}
-	else
-	{
+	/*
+	 * PQresultErrorField will return NULL if "result" is NULL, or if there is
+	 * no such field, which will happen for libpq-generated errors.  Fall back
+	 * to PQerrorMessage in such cases.
+	 */
+	sqlstate = PQresultErrorField(result, PG_DIAG_SQLSTATE);
+	if (sqlstate == NULL)
 		sqlstate = ECPG_SQLSTATE_ECPG_INTERNAL_ERROR;
+	message = PQresultErrorField(result, PG_DIAG_MESSAGE_PRIMARY);
+	if (message == NULL)
 		message = PQerrorMessage(conn);
-	}
 
 	if (strcmp(sqlstate, ECPG_SQLSTATE_ECPG_INTERNAL_ERROR) == 0)
 	{
 		/*
 		 * we might get here if the connection breaks down, so let's check for
